### PR TITLE
Bunch of improvements

### PR DESCRIPTION
**Pull request by zimbatm**
Each commit is a logical commit that improves napalm one way or the other

### CODE REVIEW

File: default.nix [hunk 1]
```diff
@@ -112,18 +113,13 @@ let
         snapshot = pkgs.writeText "npm-snapshot"
           (builtins.toJSON (snapshotFromPackageLockJson actualPackageLock));
 
-        buildInputs = [
```
Comment (nmattia) [inline default.nix]: are those really not used anywhere?

File: default.nix [hunk 2]
```diff
@@ -90,7 +90,8 @@ let
   buildPackage =
     src:
     attrs@
-    { packageLock ? null
+    { name ? "build-npm-package"
```
Comment (nmattia) [inline default.nix]: How about we default to whatever is in the lockfile? But 👍 nonetheless

File: default.nix [hunk 3]
```diff
@@ -256,28 +257,14 @@ in
...
-            echo "Dist wasn't generated"
-            exit 1
-          else
-            touch $out
-          fi
+      buildPackage sources.deckdeckgo-starter {
+        name = "deckdeckgo-starter";
+        npmCommands = [ "npm install" "npm run build" ];
+        installPhase = ''
+          mv dist $out
```
Comment (nmattia) [inline default.nix]: let's keep the check for `index.html`!

File: default.nix [hunk 4]
```diff
@@ -122,9 +122,27 @@ let
...
+            # will produce 3 parts e.g.
+            # "@someorg/somepackage" -> [ "@someorg/" "someorg" "somepackage" ]
+            # "somepackage" -> [ null null "somepackage" ]
+            parts = builtins.tail (builtins.match "^(@([^/]+)/)?([^/]+)$" pname);
+            # if there is no organisation we need to filter out null values.
+            non-null = builtins.filter (x: x != null) parts;
+          in builtins.concatStringsSep "-" non-null;
+
+        # NOTE: this is an import-from-derivation
+        packageJSON = pkgs.lib.importJSON "${toString src}/package.json";
```
Comment (nmattia) [inline default.nix]: What about `importJSON (src + "/package.json")`? also I don't think this is always IFD, but only when `src` is not a `path`.

moreover, it'd be nice to do a `pathExists` and, if false, then throw a nice error!

File: default.nix (hunk 2 above)
Comment (zimbatm) [reply default.nix]: Yeah that happens in a later commit, see ee32bd5e1840214601c7cd206d336aee28919df0

File: default.nix (hunk 1 above)
Comment (zimbatm) [reply default.nix]: `buildInputs` was only used to be combined and define `newBuildInputs`, which is then used below as the `buildInputs` of the derivation :)

File: default.nix (hunk 4 above)
Comment (zimbatm) [reply default.nix]: good points, I will see if I can improve that.

File: default.nix (hunk 3 above)
Comment (zimbatm) [reply default.nix]: fixed in 23035b243e972893aec33d1ab814b8ed2367d0f8

File: default.nix (hunk 4 above)
Comment (zimbatm) [reply default.nix]: sort of fixed in 502dee9 . It's still falling back on the defaults but emits a warning instead if the package.json is missing.

File: default.nix [hunk 5]
```diff
@@ -19,68 +16,83 @@ with rec
...
+      if hasPackageLock then src + "/package-lock.json"
+      else if hasNpmShrinkwrap then src + "/npm-shrinkwrap.json"
+      else null;
+
+  # Returns the package.json as nix values. If not found, returns an empty
+  # attrset.
+  readPackageJSON = src:
+    let
+      toplevel = builtins.readDir src;
+      hasPackageJSON = builtins.hasAttr "package.json" toplevel;
```
Comment (nmattia) [inline default.nix]: Why not `builtins.pathExists`?

Comment (zimbatm) [reply default.nix]: I just copied what you had done with `findPackageLock`.

**Comment by zimbatm** [comment]
Shall we merge for now? There is still much more that needs to be done I think but one thing at a time :)

**Comment by nmattia** [comment]
Let's!

**Pull request merged**